### DISCUSSION TITLE
Can someone point me to where the strength is computed?

### DISCUSSION THREAD

**Discussion by MatthewChatham** [iSoron/uhabits#580] [Technical Discussion]
From reading other issues and the FAQs I understand the app uses exponential smoothing. I'm a data nerd, so out of pure curiosity would anyone be able to point me to the bit of code that calculates the score?

**Comment by NellyWhads** [comment]
I think [this is what you're looking for](https://github.com/iSoron/uhabits/blob/dev/android/uhabits-core/src/main/java/org/isoron/uhabits/core/models/ScoreList.java#L264)!

**Comment by iSoron** [comment]
[This is also relevant](https://github.com/iSoron/uhabits/blob/aee0da2c649aae5bfae38e587858a3e34fc5414e/android/uhabits-core/src/main/java/org/isoron/uhabits/core/models/Score.java#L49). It's a very simple formula.

**Reply by naught101** [reply]
No much of a Java fan, but damn, that's some pretty code.

**Comment by Erudition** [comment]
```
     * The frequency of the habit is the number of repetitions divided by the
     * length of the interval. For example, a habit that should be repeated 3
     * times in 8 days has frequency 3.0 / 8.0 = 0.375.
```
Does this mean that completing the habit _more_ than 3 times in an 8 day interval would exceed the 0.375 multiplier, thus artificially boosting your score? Since I expect the score is supposed to be about "how reliably do I satisfy the prescribed frequency" rather than "how frequently do I complete the habit", the score should not increase after the third completion in an 8-day period.

Update: Just tested with a new habit, indeed it does work as expected. Score goes up to 15% and stays there after 3 checks.
However, with any more than 3 checkmarks (4-8, starting 8 days ago) the graph shows nonzero data for the day before any checkmarks were ever added - a spike which drops down to zero on day 1. Hopefully just a weird caching issue from unchecking things.

**Reply by iSoron** [reply]
In some previous versions of the app, completing the habit more often than scheduled could make your score increase to 100% very quickly. In newer versions of the app, this is no longer possible. Performing a weekly habit twice per week doesn't make your score improve any faster than performing it once per week.

If you think the app is not calculating scores correctly, [feel free to submit a detailed bug report](https://github.com/iSoron/uhabits/issues/new?assignees=&labels=&template=bug_report.md&title=). Screenshots are particularly helpful.
